### PR TITLE
Implement SA1516 diagnostic, code fix and tests

### LINKED ISSUE #117: SA1516: ElementsMustBeSeparatedByBlankLine
http://www.stylecop.com/docs/SA1516.html

### PR DESCRIPTION
**Pull request by pdelvo**
Please have a look at this. This one was a little bit tricky. Fixes #117

### CODE REVIEW

**Comment by sharwell** [comment]
This warning is getting reported for `static readonly` fields which are not separated by spaces. I'm not sure what the solution is, but the following items need to be addressed:

1. The entire `FieldDeclarationSyntax` is getting underlined, even when it spans multiple lines.
2. The associated code fix does not provide any corrections for this situation.

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/LayoutRules/SA1516UnitTests.cs [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+
+namespace StyleCop.Analyzers.Test.LayoutRules
+{
+    public class SA1516UnitTests : CodeFixVerifier
+    {
+        private const string DiagnosticId = SA1516ElementsMustBeSeparatedByBlankLine.DiagnosticId;
+
+        private const string CorrectCode = @"extern alias Foo1;
+
+using System;
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/LayoutRules/SA1516UnitTests.cs]: :exclamation: Can you update this to include multiple using directives, including a using alias directive?

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/LayoutRules/SA1516UnitTests.cs [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+using StyleCop.Analyzers.LayoutRules;
+using TestHelper;
+
+namespace StyleCop.Analyzers.Test.LayoutRules
+{
+    public class SA1516UnitTests : CodeFixVerifier
+    {
+        private const string DiagnosticId = SA1516ElementsMustBeSeparatedByBlankLine.DiagnosticId;
+
+        private const string CorrectCode = @"extern alias Foo1;
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/LayoutRules/SA1516UnitTests.cs]: :question: Does the reference StyleCop implementation require a blank line between the `extern alias` directive and the using alias directives?

Comment (pdelvo) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/LayoutRules/SA1516UnitTests.cs]: As far as I know it does

File: StyleCop.Analyzers/StyleCop.Analyzers.Test/LayoutRules/SA1516UnitTests.cs (hunk 1 above)
Comment (pdelvo) [inline StyleCop.Analyzers/StyleCop.Analyzers.Test/LayoutRules/SA1516UnitTests.cs]: I will do that tomorrow (its alomst 3am over here ;))

File: StyleCop.Analyzers/StyleCop.Analyzers/LayoutRules/SA1516CodeFixProvider.cs [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+                {
+                    context.RegisterCodeFix(CodeAction.Create("Insert new line", token => GetTransformedDocument(context, syntaxRoot, node, (SyntaxTriviaList)leadingTrivia)), diagnostic);
+                }
+            }
+        }
+
+        private static Task<Document> GetTransformedDocument(CodeFixContext context, SyntaxNode syntaxRoot, SyntaxNode node, SyntaxTriviaList leadingTrivia)
+        {
+            var newTriviaList = leadingTrivia;
+            newTriviaList = newTriviaList.Insert(0, SyntaxFactory.EndOfLine("\r\n", elastic: true));
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers/LayoutRules/SA1516CodeFixProvider.cs]: You can also use `SyntaxFactory.ElasticCarriageReturnLineFeed` here.

**Comment by vweijsters** [comment]
I'm missing the handling and testing of interfaces and indexer properties.

File: StyleCop.Analyzers/StyleCop.Analyzers/LayoutRules/SA1516ElementsMustBeSeparatedByBlankLine.cs [hunk 4]
```diff
@@ -76,7 +83,174 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
...
+                    // or the current declaration has documentation
+                    // or the current declaration is not a field declaration,
+                    if (IsMultiline(members[i - 1])
+                        || !members[i - 1].IsKind(members[i].Kind())
+                        || XmlCommentHelper.HasDocumentation(members[i])
+                        || !members[i].IsKind(SyntaxKind.FieldDeclaration))
+                    {
+                        ReportIfThereIsNoBlankLine(context, members[i - 1], members[i]);
+                    }
+                }
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers/LayoutRules/SA1516ElementsMustBeSeparatedByBlankLine.cs]: Shouldn't the documentation one be handled by SA1514?

File: StyleCop.Analyzers/StyleCop.Analyzers/LayoutRules/SA1516ElementsMustBeSeparatedByBlankLine.cs [hunk 5]
```diff
@@ -76,7 +83,174 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
...
-            // TODO: Implement analysis
+            context.RegisterSyntaxNodeAction(this.HandleTypeDeclaration, SyntaxKind.ClassDeclaration);
+            context.RegisterSyntaxNodeAction(this.HandleTypeDeclaration, SyntaxKind.StructDeclaration);
+
+            context.RegisterSyntaxNodeAction(this.HandleCompilationUnit, SyntaxKind.CompilationUnit);
+            context.RegisterSyntaxNodeAction(this.HandleNamespaceDeclaration, SyntaxKind.NamespaceDeclaration);
+
+            context.RegisterSyntaxNodeAction(this.HandlePropertyDeclaration, SyntaxKind.PropertyDeclaration);
+            context.RegisterSyntaxNodeAction(this.HandlePropertyDeclaration, SyntaxKind.EventDeclaration);
+        }
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers/LayoutRules/SA1516ElementsMustBeSeparatedByBlankLine.cs]: `RegisterSyntaxNodeAction` will accept multiple syntax kinds in a single call

File: StyleCop.Analyzers/StyleCop.Analyzers/LayoutRules/SA1516CodeFixProvider.cs (hunk 3 above)
Comment (pdelvo) [inline StyleCop.Analyzers/StyleCop.Analyzers/LayoutRules/SA1516CodeFixProvider.cs]: Will do

File: StyleCop.Analyzers/StyleCop.Analyzers/LayoutRules/SA1516ElementsMustBeSeparatedByBlankLine.cs (hunk 4 above)
Comment (pdelvo) [inline StyleCop.Analyzers/StyleCop.Analyzers/LayoutRules/SA1516ElementsMustBeSeparatedByBlankLine.cs]: I'll clean that up. That wasn't intended ;)

File: StyleCop.Analyzers/StyleCop.Analyzers/LayoutRules/SA1516ElementsMustBeSeparatedByBlankLine.cs (hunk 5 above)
Comment (pdelvo) [inline StyleCop.Analyzers/StyleCop.Analyzers/LayoutRules/SA1516ElementsMustBeSeparatedByBlankLine.cs]: In my opinion that way it is easier to read if you only register one kind each line

**Comment by pdelvo** [comment]
I'll add tests for structs and I will look into interfaces too. I'm not sure if the same rules apply there too.

**Comment by pdelvo** [comment]
I addressed your feedback

**Comment by vweijsters** [comment]
Nice work:+1:
Could you (for completeness) also add some test cases that include indexer properties?

**Comment by pdelvo** [comment]
Done!

**Pull request merged**